### PR TITLE
[#71] New rule `unused_callbacks`

### PR DESCRIPTION
**Pull request by diegomanuel**
## Issue: https://github.com/AdRoll/rebar3_hank/issues/71

The first _acceptable_ version of this rule is done, but the macro cases are still unsupported!
Issues to handle macros were created:
* https://github.com/AdRoll/rebar3_hank/issues/81
* https://github.com/AdRoll/rebar3_hank/issues/82

### CODE REVIEW

File: src/rules/unused_callbacks.erl [hunk 1]
```diff
@@ -0,0 +1,15 @@
+%% @doc A rule to detect unused callbacks.
+%%      <p>The rule emits a warning if it can't find the callback's atom name
+%%      being used within a module. It will NOT emit a warning if an atom named
+%%      as a callback is being used, no matter for what that atom is used.</p>
```
Comment (elbrujohalcon) [inline src/rules/unused_callbacks.erl:4]: Since the docs are used as informative descriptions of each rule, I think we should clearly state the goal of the rule before diving into the implementation details and the caveats of _how_ it works.
In other words, I would add something along the lines of…

> This rule will check all callbacks defined in a module and find those that are not used anywhere in the module itself. The assumption is that if you define a callback for a behavior, your _generic module_ (where the callback is defined) should call that function at some point, using the implementation provided by the specific module (the one that implements the behavior).

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]

File: src/rules/unused_callbacks.erl (hunk 1 above)
Comment (diegomanuel) [reply src/rules/unused_callbacks.erl]: Did the explanation adjustment, check-n-tell!

File: priv/test_files/unused_callbacks/warnings.erl [hunk 2]
```diff
@@ -0,0 +1,15 @@
+-module(warnings).
+
+-callback unused_callback(map()) -> boolean().
+-callback used_callback(map()) -> boolean().
+
+-export([unused_callback/1, used_callback/1, call_callback/1]).
+
+unused_callback(#{module := Module, state := State}) ->
```
Comment (pbrudnick) [inline priv/test_files/unused_callbacks/warnings.erl:8]: it's used here I think

File: priv/test_files/unused_callbacks/clean.erl [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+-callback used_callback(map()) -> boolean().
+-callback used_atom(atom()) -> list().
+
+-export([used_callback/1, used_atom/1]).
+
+used_callback(#{module := Module, state := State}) ->
+    Module:used_callback(State).
+
+used_atom(Module) ->
+    other_module:call(used_atom, Module).
```
Comment (elbrujohalcon) [inline priv/test_files/unused_callbacks/clean.erl:12]: To make it _clearer_… let's use…
```suggestion
    erlang:apply(Module, used_atom, [another_atom]).
```
Since that's a perfectly valid way of using the callback that _is_ captured by Hank.

File: priv/test_files/unused_callbacks/warnings.erl (hunk 2 above)
Comment (elbrujohalcon) [reply priv/test_files/unused_callbacks/warnings.erl:8]: The function name is unfortunate… but it's a great test since Hank shouldn't consider a function with the same name as a callback as usage. The callback is definitely *not* used here.

File: priv/test_files/unused_callbacks/warnings.erl [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+%% nothing to do with the callback itself like in this example
+%% in where `other_module:call/2` is "opaque" to us and its
+%% implementation may not use this module callbacks at all
+unused_callback(#{module := Module, state := State}) ->
+    other_module:call(Module, State).
+
+%% This function counts as callback usage (although it really isn't)
+%% because it's being called from `call_callback/1` and Hank detects
+%% the atom `used_callback` in it, although `other_module:do/1`
+%% implementation may not use this module callbacks at all
```
Comment (elbrujohalcon) [inline priv/test_files/unused_callbacks/warnings.erl:20]: This is not _exactly_ the truth. What counts as _usage_ is not the function itself. What counts as usage is the invocation, within `call_callback/1`. I would remove this comment and, if you really have to, add a similar but more precise comment to `call_callback/1`.
But… This is a justification of an edge case which is a _false negative_ (i.e. a callback that is unused but Hank can't detect). Personally, I would not put this in a test (i.e. I would not validate that Hank _can't_ find something). Or… if you really really want to validate that… I would put the comment on the callback itself (line 4 here) with something like…

```erlang
%% @todo [#…] Detect that this callback is actually unused.
-callback used_callback(…
```

File: src/rules/unused_callbacks.erl [hunk 5]
```diff
@@ -0,0 +1,64 @@
+%% @doc A rule to detect unused callbacks.
+%%      <p>This rule will check all callbacks defined in a module and find
+%%      those that are not used anywhere in the module itself.</p>
+%%      <p>It will emit a warning if it can't find the callback's atom name
+%%      being used anywhere within a module. It will NOT emit a warning if an atom
+%%      named as a callback is being used, no matter for what that atom is used.</p>
```
Comment (elbrujohalcon) [inline src/rules/unused_callbacks.erl:6]: Maybe add something like _This limitation is due to the fact that there are many ways to call a function in Erlang (particularly when dynamic calls are involved)._ to explain that it's not the _desired_ behavior but it is what what we were _able_ to implement.

File: priv/test_files/unused_callbacks/warnings.erl (hunk 4 above)
Comment (elbrujohalcon) [reply priv/test_files/unused_callbacks/warnings.erl:20]: But to be clear: I'm 100% sure that we'll never be able to get rid of that `@todo` while maintaining dialyzer-levels of certainty™️.

File: src/rules/unused_callbacks.erl [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    % ct:print("CALLBACKS: ~p", [Callbacks]),
+    analyze_callbacks(AST, Callbacks).
+
+analyze_callbacks(_AST, []) ->
+    []; %% Skip files with no callback definitions
+analyze_callbacks(AST, Callbacks) ->
+    Functions = [Node || Node <- AST, erl_syntax:type(Node) == function],
+    [Result
+     || {File, Line, Callback} <- Callbacks,
+        Result <- analyze_callback(File, Line, Callback, Functions)].
```
Comment (elbrujohalcon) [inline src/rules/unused_callbacks.erl:45]: Why don't you turn `analyze_callback/4` into a boolean function and then…
```suggestion
    [set_result(File, Line, Callback)
     || {File, Line, Callback} <- Callbacks,
        not is_used_callback(Callback, Functions)].
```

File: src/rules/unused_callbacks.erl [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+analyze_callback(File, Line, Callback, Functions) ->
+    case lists:any(fun(Function) -> function_has_atom(Function, Callback) end, Functions) of
+        true ->
+            [];
+        false ->
+            [set_result(File, Line, Callback)]
+    end.
```
Comment (elbrujohalcon) [inline src/rules/unused_callbacks.erl:53]: If you follow my previous comment, this function can end up being as simple as…
```suggestion
is_used_callback(Callback, Functions) ->
    lists:any(fun(Function) -> function_has_atom(Function, Callback) end, Functions).
```

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
It's looking better… I left a few tips here and there.
One thing you need to do is add tests with macros… macros everywhere!!

```erlang
-callback used_callback() -> this | one | is | used | through | the | macro.
-callback unused_callback() -> this | one | isnt | used.

-define(unused_callback, {this, macro, doesnt, use, the, callback}).

-define(CALLBACK, used_callback).
using_the_callback(M) -> M:?CALLBACK().

not_using_the_callback() -> ?unused_callback.
```

File: priv/test_files/unused_callbacks/warnings.erl (hunk 4 above)
Comment (diegomanuel) [reply priv/test_files/unused_callbacks/warnings.erl:20]: Removed the comment to avoid confusion!

**Comment by diegomanuel** [comment]
The first _acceptable_ version of this rule is done, but the macro cases are still unsupported!
Issues to handle macros were created:
* https://github.com/AdRoll/rebar3_hank/issues/81
* https://github.com/AdRoll/rebar3_hank/issues/82

File: priv/test_files/unused_callbacks/macros.erl [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+-define(unused_callback, {this, macro, doesnt, use, the, callback}).
+-define(CALLBACK, used_callback).
+
+-export([using_the_callback/1, not_using_the_callback/0]).
+
+using_the_callback(Module) ->
+    Module:?CALLBACK().
+
+not_using_the_callback() ->
+    ?unused_callback.
```
Comment (pbrudnick) [inline priv/test_files/unused_callbacks/macros.erl:15]:
```suggestion
    ?unused_callback.
    
```

**Review by pbrudnick — APPROVED** [review]

File: src/rules/unused_callbacks.erl [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+%%      <p>The assumption is that if you define a callback for a behavior,
+%%      your generic module (where the callback is defined) should call
+%%      that function at some point, using the implementation provided
+%%      by the specific module (the one that implements the behavior).</p>
+%%      <p>To avoid this warning, remove the unused callback definition.</p>
+-module(unused_callbacks).
+
+-behaviour(hank_rule).
+
+%% @TODO [#81 + #82] Correctly handle macros
```
Comment (elbrujohalcon) [inline src/rules/unused_callbacks.erl:18]: For edoc to read this properly, this should be placed above `-module(…)` and it should be written as `@todo` (i.e. lowercase).

File: src/rules/unused_callbacks.erl [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    [Result || {File, AST} <- FilesAndASTs, Result <- analyze_file(File, AST)].
+
+analyze_file(File, AST) ->
+    CbNodes = [Node || Node <- AST, hank_utils:node_has_attrs(Node, [callback])],
+    Callbacks =
+        lists:map(fun(CbNode) ->
+                     [CbDataArgs | _] = erl_syntax:attribute_arguments(CbNode),
+                     [CbDataTuple | _] = erl_syntax:tuple_elements(CbDataArgs),
+                     [CbName, CBArit] = erl_syntax:tuple_elements(CbDataTuple),
+                     {File,
```
Comment (elbrujohalcon) [inline src/rules/unused_callbacks.erl:33]: You don't need to put `File` in every tuple. You can just add `File` as a first argument on `analyze_callbacks/2` (i.e. `analyze_callbacks(File, AST, Callbacks).`

File: src/rules/unused_callbacks.erl [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+is_used_callback(Callback, Functions) ->
+    lists:any(fun(Function) -> function_has_atom(Function, Callback) end, Functions).
+
+function_has_atom(FuncNode, FuncName) ->
+    FuncBodies =
+        [Body
+         || Clause <- erl_syntax:function_clauses(FuncNode),
+            Body <- erl_syntax:clause_body(Clause)],
+    FuncAtoms = hank_utils:node_atoms(FuncBodies),
+    lists:any(fun(FuncAtom) -> FuncAtom =:= FuncName end, FuncAtoms).
```
Comment (elbrujohalcon) [inline src/rules/unused_callbacks.erl:57]:
```suggestion
    lists:member(FuncName, FuncAtoms).
```

File: test/unused_callbacks_SUITE.erl [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+    File = "warnings.erl",
+    [#{file := File,
+       line := 3,
+       text := <<"Callback unused_callback/1 is not used anywhere in the module">>}] =
+        analyze([File]),
+    ok.
+
+%% @doc Hank finds unused callbacks with macros
+%% @TODO [#81 + #82] Correctly handle macros
```
Comment (elbrujohalcon) [inline test/unused_callbacks_SUITE.erl:27]:
```suggestion
%% @todo [#81 + #82] Correctly handle macros
```

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
This looks great! Just a few minor tweaks…

**Review by elbrujohalcon — APPROVED** [review]
